tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+        ) in FindUndocumentedSymbols.documentable_symbol_patterns.values():
+            matches = [
+                m for m in each_matching(root, pat) if not check_func or check_func(*m)
+            ]
+            for node, _ in matches:
+                yield node
+
+    def has_doc_comment(self, node: dyno.AstNode) -> bool:
+        """
+        This node has a doc comment if
+        - the previous node is a comment
+        """
+
+        if (
+            (prev := self._get_previous_sibling(node))
+            and isinstance(prev, dyno.Comment)
+            and is_docstring_comment(prev)
+        ):
+            return True
+
+        return False
+
+    def __call__(self):
+        syms = [s for s in self.get_documentable_symbols([a for a in self.ast])]
+        for s in syms:
+            if not is_nodoc(s) and not self.has_doc_comment(s):
+                if self.ignore_deprecated and is_deprecated(s):
+                    continue
+                if self.ignore_unstable and is_unstable(s):
+                    continue
+                yield s
+
+
+def get_files(files: List[str], check_suffix=False) -> Generator:
+    """
+    Yield all chapel files, following directories recursively
+    """
+
+    for file in files:
+        if os.path.isfile(file):
+            # just yield any file given if not checking the suffix
+            if not check_suffix:
+                yield file
+            elif check_suffix and os.path.splitext(file)[1] == ".chpl":
+                yield file
+
+        elif os.path.isdir(file):
+            for root, subdirs, files in os.walk(file):
+                # yield files in directory files, always check the suffix
+                yield from get_files([os.path.join(root, file) for file in files], True)
+                # recurse into sub dirs, always checl the suffix
+                yield from get_files([os.path.join(root, d) for d in subdirs], True)
+        else:
+            print("Error: path was not file or directory", file=sys.stderr)
+            exit(1)
+
+
+def get_trees(filenames: List[str]) -> Generator:
+    """
+    Yield dyno asts and their filename.
+    Some files might have the same name, which Dyno really doesn't like.
+    Strateify files into "buckets"; within each bucket, all filenames are
+    unique. Between each bucket, re-create the Dyno context to avoid giving
+    it complicting files.
+    """
+    basenames = defaultdict(lambda: 0)
+    buckets = defaultdict(lambda: [])
+    for filename in filenames:
+        filename_expanded = os.path.realpath(os.path.expandvars(filename))
+
+        basename = os.path.basename(filename_expanded)
+        bucket = basenames[basename]
+        basenames[basename] += 1
+        buckets[bucket].append((filename, filename_expanded))
+
+    for filenames in buckets.values():
+        ctx = dyno.Context()
+        for filename, filename_expanded in filenames:
+            ast = ctx.parse(filename_expanded)
+            yield (filename, ast)
```

Comment:
Perhaps you can wait for the chplcheck PR to go in, which moves this whole buckets business into a library function?

Reply:
The one thing that is different here is the I also yield the original filename, as it makes for prettier printing. There are other ways around that problem though, and I agree that the duplication would be bad